### ISSUE TITLE
Helm provider does not trigger an update, if local templates change

### ISSUE THREAD

**Issue by markusheiden** [hashicorp/terraform-provider-helm#821]
## Terraform version, Kubernetes provider version and Kubernetes version
```
Terraform version: 1.1.3
Helm Provider version: 2.4.1
Kubernetes version: 1.20
```
## Terraform configuration
```hcl
resource "helm_release" "myapp" {
  namespace = "namespace"
  name      = "myapp"
  chart     = "${path.module}"
}
```

## Question
```
Why does the Helm provider NOT trigger an update, if the files in the "templates" directory change? 
How to achieve that?

My temporary solution is to compute a hash of the templates files and use that as a dummy value.
```

```hcl
resource "helm_release" "myapp" {
  ...

  // Trigger updates on template changes.
  set {
    name = "templatesHash"
    value = module.templates_hash.hash
  }
}

module "templates_hash" {
  source = "../helm-hash"
  chart_directory = path.module
}
```

helm-hash module:
```hcl
variable "chart_directory" {
  description = "Chart directory"
  type        = string
}

locals {
  templates_directory = "${var.chart_directory}/templates"
  template_hashes     = {
  for path in sort(fileset(local.templates_directory, "**")) :
  path => filebase64sha512("${local.templates_directory}/${path}")
  }
  hash                = base64sha512(jsonencode(local.template_hashes))
}

output "hash" {
  value = local.hash
}

```

**Comment by junzebao** [comment]
We noticed the same issue. What we do is to bump the chart version every time there's a change in the templates. That'll trigger a helm upgrade.

**Comment by markusheiden** [comment]
I am aware of that, but the risk of not updating changed parts due to a forgotten version increase is not acceptable for me. It would be nice if a solution similar to my hack would make it into the provider.

**Comment by vikaskoppineedi** [comment]
I am facing the same issue, I am planning to add triggers with timestamp to overcome this issue, but this will cause the deployment to trigger everytime

**Comment by beepdot** [comment]
> We noticed the same issue. What we do is to bump the chart version every time there's a change in the templates. That'll trigger a helm upgrade.

Thanks @junzebao. It works well and also is something very logical to do - _just bump the chart version on chart changes_

**Comment by Oliniusz** [comment]
For me, bumping the chart version by hand sounds like: _we've automated everything, we just need to trigger the automation by hand each time_

I'm looking for a solution that will trigger `helm_release` by any changes done under the `postrender` block and I think the checksum is the only way I know of at the moment. I calculate the changes in a directory and use it to change the description:


```terraform
resource "helm_release" "istio-ingressgateway-internal" {
  name        = "istio-ingressgateway-internal"
  repository  = "https://istio-release.storage.googleapis.com/charts"
  chart       = "gateway"
  version     = "1.17.2"
  description = sha1(join("", [for f in sort(fileset("${path.module}/helm-values/istio-ingressgateway-internal-kustomize", "*")) : filesha1("${path.module}/helm-values/istio-ingressgateway-internal-kustomize/${f}")]))
  namespace   = kubernetes_namespace.istio-ingress.metadata[0].name
  timeout     = "900"
  values = [
    templatefile("${path.module}/helm-values/istio-ingressgateway-internal.yaml", { infra_name = var.infra_name, eks_suffix = var.eks_suffix })
  ]
  postrender {
    binary_path = "${path.module}/helm-values/istio-ingressgateway-internal-kustomize/kustomize.sh"
  }
  depends_on = [
    helm_release.istio-istiod,
    kubernetes_namespace.istio-ingress
  ]
}
```

**Comment by djmcgreal-cc** [comment]
I'm surprised this hasn't had more attention! Is there an explanation somewhere why this can't happen? It's really counterintuitive, and so even with work-arounds, new helm charts created aren't always spotted as requiring extra work to force a deployment.

`helm template | kubectl diff -f -` seems fairly straightforward, though I guess it'd need to filter hooks.

**Comment by PowerSurj** [comment]
description field changes cause provider errors
i'm on terraform v1.6.4

```
Error: Provider produced inconsistent final plan
When expanding the plan for helm_release.myapp[0] to include new values learned so far during apply, provider "registry.terraform.io/hashicorp/helm" produced an invalid new value for .description: was cty.StringVal("4906e97bd629cbf998cd33647ab879d8ef90536e"), but now cty.StringVal("5380c9793e0f8669c17f5e45ee25ab927fe9bb28").

This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

**Comment by alfechner** [comment]
We set `reset_values  = true` and that works fine for us.

My understanding is that the default behaviour of Helm is to [consider this flag `true` if it's not set](https://medium.com/@kcatstack/understand-helm-upgrade-flags-reset-values-reuse-values-6e58ac8f127e). That's the behaviour we're all used to.

The Terraform `helm_release` resource however [defaults to `false`](https://registry.terraform.io/providers/hashicorp/helm/latest/docs/resources/release#reset_values). That explains the unusual behaviour I guess.

**Comment by PowerSurj** [comment]
found a workaround that worked for me which is through the set values
https://github.com/hashicorp/terraform-provider-helm/issues/769
this worked for me

**Comment by PowerSurj** [comment]
The cause of error seems to have been coming due to the Chart with dependencies.
Description driven updates work fine without `reset_values` and with Charts that have no dependencies.
My solution is to break the dependencies into independent charts.

**Comment by ghost** [comment]
> The cause of error seems to have been coming due to the Chart with dependencies.

I have a local-only chart without any dependencies. When I'm modifying a template in a local folder, Terraform does not detect any changes.

**Comment by alfechner** [comment]
@PowerSurj what's the issue with using `reset_values`?

**Comment by lipika-pal-partior** [comment]
> ```terraform
> output "hash" {
>   value = local.hash
> }
> ```

reset_values will reset it to the default values of the chart but if you want to use custom values this will not work.

**Comment by lipika-pal-partior** [comment]
> @PowerSurj what's the issue with using `reset_values`?

reset_values will reset the values to the default ones built into the chart... so if you wish to use custom values it will be over-written

**Comment by alfechner** [comment]
@lipika-pal-partior you refer to [set](https://registry.terraform.io/providers/hashicorp/helm/latest/docs/resources/release#set), [set_list](https://registry.terraform.io/providers/hashicorp/helm/latest/docs/resources/release#set_list) and [set_sensitive](https://registry.terraform.io/providers/hashicorp/helm/latest/docs/resources/release#set_sensitive), right?

I wasn't aware of those arguments, thanks for pointing me into that direction.

But wouldn't the custom values are applied back the moment terraform is applied? So the `reset_values` would reset them and then they would be added back?

**Comment by djmcgreal-cc** [comment]
`reset_values` doesn't work if the post_renderer executable changes. This really needs a more robust solution.

**Comment by oubaydos** [comment]
any updates ?
I personally use this workaround 
```tf
set {
    name  = "ignoreMe"
    value = timestamp()
  }
  ```
  to trigger the deployment each time and let helm control which templates/values have changed

**Comment by hubertlepicki** [comment]
FYI, this is related to chart version. When you modify the chart/templates, edit Chart.yaml and bump up the version variable so that the changes get picked up and applied by Terraform.

**Comment by NeodymiumFerBore** [comment]
Using helm provider 2.15.0, I'm using the following workaround. It computes the sha1sum of all templates files, concatenate them, compute the sha1sum of the resulting string, and set it to a dummy Value. It does not trigger a change every time like `timestamp` does.

```hcl
resource "helm_release" "mychart" {
  name = "myapp"
  chart = "${path.module}/charts/mychart"
  set {
    name  = "templates_hash"
    value = sha1(join("", [for f in fileset("${path.module}/charts/mychart/templates", "*") : filesha1("${path.module}/charts/mychart/templates/${f}")]))
  }
}
```

**Comment by jeremymcgee73** [comment]
I do not believe the workaround is working with `3.0.0-pre1`. I reverted to `2.17.0` and this started working again as expected.
